File: Data.Resumption/Execution/ResponseCache.cs
using System;
using System.Collections.Generic;

namespace Data.Resumption.Execution
{
    internal class ResponseCache
    {
        private readonly DataSourceCache _nullDataSource = new DataSourceCache(null);
        private readonly Dictionary<IComparable, DataSourceCache> _dataSources
            = new Dictionary<IComparable, DataSourceCache>();

        public void Store(IComparable dataSource, IComparable identity, SuccessOrException response)
        {
            if (dataSource == null)
            {
                _nullDataSource.Store(identity, response);
            }
            else
            {
                DataSourceCache dataSourceCache;
                if (!_dataSources.TryGetValue(dataSource, out dataSourceCache))
                {
                    dataSourceCache = new DataSourceCache(dataSource);
                    _dataSources[dataSource] = dataSourceCache;
                }
                dataSourceCache.Store(identity, response);
            }
        }

        public void Invalidate(IComparable dataSource)
        {
            if (dataSource == null)
            {
                _nullDataSource.Clear();
            }
            else
            {
                _dataSources.Remove(dataSource);
            }
        }

        public SuccessOrException? CheckCache(IComparable dataSource, IComparable identity)
        {
            if (dataSource == null)
            {
                return _nullDataSource.CheckCache(identity);
            }
            DataSourceCache dataSourceCache;
            return _dataSources.TryGetValue(dataSource, out dataSourceCache)
                ? dataSourceCache.CheckCache(identity)
                : null;
        }
    }
}

File: Rezoom/ResponseCache.fs
namespace Rezoom
open System.Collections.Generic

type private Response = obj
type private DataSource = obj
type private Identity = obj

type ResponseCache() =
    let nullDataSource = new Dictionary<Identity, Response>()
    let byDataSource = new Dictionary<DataSource, Dictionary<Identity, Response>>()

    member __.Invalidate(dataSource : DataSource) =
        if isNull dataSource then nullDataSource.Clear()
        else ignore <| byDataSource.Remove(dataSource)

    member __.Store(dataSource : DataSource, identity : Identity, value : Response) =
        if isNull dataSource then
            nullDataSource.[identity] <- value
        else
            let mutable subCache : Dictionary<Identity, Response> = null
            if not <| byDataSource.TryGetValue(dataSource, &subCache) then
                subCache <- new _()
                byDataSource.[dataSource] <- subCache
            subCache.[identity] <- value

    member __.TryGetValue(dataSource : DataSource, identity : Identity, value : Response byref) =
        if isNull dataSource then
            nullDataSource.TryGetValue(identity, &value)
        else
            let mutable subCache : Dictionary<Identity, Response> = null
            if byDataSource.TryGetValue(dataSource, &subCache) then
                subCache.TryGetValue(identity, &value)
            else false